x/evm/keeper/keeper.go

```diff
@@ -281,107 +193,253 @@ func (k Keeper) GetLogSizeTransient() uint64 {
+		coins := sdk.NewCoins(sdk.NewCoin(params.EvmDenom, sdk.NewIntFromBigInt(new(big.Int).Neg(delta))))
+		if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, cosmosAddr, types.ModuleName, coins); err != nil {
+			return err
+		}
+		if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, coins); err != nil {
+			return err
+		}
+	default:
+		// not changed
+	}
+	return nil
 }
 
-// ResetAccount removes the code, storage state, but keep all the native tokens stored
-// with the given address.
-func (k Keeper) ResetAccount(addr common.Address) {
-	k.DeleteCode(addr)
-	k.DeleteAccountStorage(addr)
+// SetAccount updates nonce/balance/codeHash together.
+func (k *Keeper) SetAccount(ctx sdk.Context, addr common.Address, account statedb.Account) error {
+	// update account
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	acct := k.accountKeeper.GetAccount(ctx, cosmosAddr)
+	if acct == nil {
+		acct = k.accountKeeper.NewAccountWithAddress(ctx, cosmosAddr)
+	}
+	ethAcct, ok := acct.(*ethermint.EthAccount)
+	if !ok {
+		return errors.New("not EthAccount")
+	}
+	if err := ethAcct.SetSequence(account.Nonce); err != nil {
+		return err
+	}
+	ethAcct.CodeHash = common.Bytes2Hex(account.CodeHash)
+	k.accountKeeper.SetAccount(ctx, ethAcct)
+
+	return k.SetBalance(ctx, addr, account.Balance)
 }
 
-// SetHooks sets the hooks for the EVM module
-func (k *Keeper) SetHooks(eh types.EvmHooks) *Keeper {
-	if k.hooks != nil {
-		panic("cannot set evm hooks twice")
+func (k *Keeper) SetState(ctx sdk.Context, addr common.Address, key, value common.Hash) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.AddressStoragePrefix(addr))
+	if len(value) == 0 {
+		store.Delete(key.Bytes())
+	} else {
+		store.Set(key.Bytes(), value.Bytes())
 	}
+}
 
-	k.hooks = eh
-	return k
+func (k *Keeper) SetCode(ctx sdk.Context, codeHash []byte, code []byte) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixCode)
+
+	// store or delete code
+	if len(code) == 0 {
+		store.Delete(codeHash)
+	} else {
+		store.Set(codeHash, code)
+	}
 }
 
-// PostTxProcessing delegate the call to the hooks. If no hook has been registered, this function returns with a `nil` error
-func (k *Keeper) PostTxProcessing(txHash common.Hash, logs []*ethtypes.Log) error {
-	if k.hooks == nil {
+// PurgeAccount handles contract's suicide call:
+// - clear balance
+// - remove code
+// - remove states
+func (k *Keeper) PurgeAccount(ctx sdk.Context, addr common.Address) error {
+	acct, err := k.GetAccount(ctx, addr)
+	if err != nil {
+		return err
+	}
+
+	if acct == nil {
 		return nil
 	}
-	return k.hooks.PostTxProcessing(k.Ctx(), txHash, logs)
-}
 
-// Tracer return a default vm.Tracer based on current keeper state
-func (k Keeper) Tracer(msg core.Message, ethCfg *params.ChainConfig) vm.Tracer {
-	return types.NewTracer(k.tracer, msg, ethCfg, k.Ctx().BlockHeight(), k.debug)
+	// clear balance
+	if acct.Balance.Sign() != 0 {
+		if err := k.SetBalance(ctx, addr, new(big.Int)); err != nil {
```

Comment:
If we clear the contract balances. Will it be burned or just disappear directly? How does it affect the token supply?

Reply:
It works like:
```
old := getBalance()
if new > old {
    mint(new - old)
} else {
    burn(old - new)
}
```

Reply:
Sorry, I mean 
In this line of `PurgeAccount`, we clear the account balance. So the removed balance will be removed from the entire token supply?

Reply:
> Sorry, I mean In this line of `PurgeAccount`, we clear the account balance. So the removed balance will be removed from the entire token supply?

That'll be handled by higher-level logic in the EVM, for example, [`selfdestruct` opcode](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L792) will do `AddBalance(beneficiary)` and `Suicide(contract_address)` together.